### PR TITLE
add FilterByTime functionality

### PR DESCRIPTION
**Pull request by KyleQianliMa**
### Description of work
This PR adds the function of "FilterbyTime to the LoadEventAsWorkspace2D function by introducing two optional input fields: "FilterByTimeStart" and "FilterByTimeStop". If both are left empty, it will by default not apply any filter. If either start or stop is entered, it will filter from start time till end or start till end time. 

The functionality is implemented by introducing "BankPulseTime" and "PulseIndexer". Once "FilterByTimeStart" and/or "FilterByTimeStop" are introduced, a TimeROI is created from BankPulseTime, then this TimeROI is used to extract event_index using PulseIndexer, then a for loop is used to loop through the filtered event_index to generate "Y" of workspace.

Related to [story 6294](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=6294)

#### Summary of work

Add FilterByTime functionality similar to the workflow of LoadEventNexus(FilterByTime) + Integeration.

Fixes #xxxx. 

### To test:

Run python scripts as below:
ws1 = LoadEventAsWorkspace2D(Filename = "BSS_11841_event.nxs", OutputWorkspace = "outputWS",
                        FilterByTimeStart = 0.0, FilterByTimeStop = 5.0, 
                        XCenter = 1.54, XWidth = 0.1,
                        Units = "Wavelength"
)

ws2 = LoadEventNexus(Filename = "BSS_11841_event.nxs", OutputWorkspace ="outputWS2",
                FilterByTimeStart = "0.0", FilterByTimeStop = "5.0")

ws3 = Integration(InputWorkspace = "outputWS2", RangeLower = 0.0, OutputWorkspace = "outputWS3")

ws3.getAxis(0).setUnit("Wavelength")
xBins = [1.463, 1.617]
for i in range(ws3.getNumberHistograms()):
    ws3.setX(i, xBins)
CompareWorkspaces(ws1,ws3)

Alternatively:
In Mantid, open LoadEventAsWorkspace2D, load mantid/build/ExternalData/Testing/Data/UnitTest/BSS_11841_event.nxs. Set FilteByTimeStart and FilterByTimeStop to specific values(i.e. 0 and 5), set XCenter to 1.54, XWidth to 0.1, units = Wavelength and generate workspace. 

Then open LoadEventNexus, load same .nxs file, set FilterByTimeStart/FilterByTimeStop to same values as before and create another workspace. Then use Integration to integrate this workspace, set RangeLower = 0 and generate workspace. Compare the Y values of the workspace. This can be done by plotting the data or look at the data specifically.

Another wayt to test is follow the LoadEventAsWorkspace2DTest.h to directly call LoadEventAsWorkspace2D, LoadEventNexus, Integrate from command line.

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Review by jmborr — CHANGES_REQUESTED** [review]
in the pull request description, can you write a manual test I can run in the workbench?

**Comment by KyleQianliMa** [comment]
> in the pull request description, can you write a manual test I can run in the workbench?
I wrote that in the wrong place and it got commented out. Should I also write something that can be directly exectued from commandline of workbench?

**Comment by jmborr** [comment]
> > in the pull request description, can you write a manual test I can run in the workbench?
> > I wrote that in the wrong place and it got commented out. Should I also write something that can be directly exectued from commandline of workbench?

Yes, a few python statements that I could execute in the workbench would be best

**Comment by KyleQianliMa** [comment]
I've created something like this:
ws1 = LoadEventAsWorkspace2D(Filename = "BSS_11841_event.nxs", OutputWorkspace = "outputWS",
                        FilterByTimeStart = 0.0, FilterByTimeStop = 5.0, 
                        XCenter = 1.54, XWidth = 0.1,
                        Units = "Wavelength"
)

ws2 = LoadEventNexus(Filename = "BSS_11841_event.nxs", OutputWorkspace ="outputWS2",
                FilterByTimeStart = "0.0", FilterByTimeStop = "5.0")

ws3 = Integration(InputWorkspace = "outputWS2", RangeLower = 0.0, OutputWorkspace = "outputWS3")

The data of ws1 and ws3 is the same. But CompareWorkspace returns False because their X-axis is different. In the c++ code, setting x-axis can be achieved by:
    // set the expected X-axis
    outputWS3->getAxis(0)->setUnit("Wavelength");
    const auto xBins = {1.463, 1.617};
    const auto histX = Mantid::Kernel::make_cow<Mantid::HistogramData::HistogramX>(xBins);
    for (size_t i = 0; i < outputWS3->getNumberHistograms(); i++)
      outputWS3->setSharedX(i, histX);
But I don't know how to generate histX with python.

**Review by jmborr — COMMENTED** [review]
This script pass
```
from mantid.simpleapi import HFIRSANS2Wavelength, LoadEventAsWorkspace2D, LoadEventNexus
filename="/HFIR/CG3/IPTS-23782/nexus/CG3_4829.nxs.h5"

LoadEventNexus(
    Filename=filename,
    OutputWorkspace="CG3_4829_1",
    FilterByTimeStart=60,
    FilterByTimeStop=120
)
HFIRSANS2Wavelength(InputWorkspace="CG3_4829_1",  OutputWorkspace="CG3_4829_1")


LoadEventAsWorkspace2D(
    Filename=filename,
    OutputWorkspace="CG3_4829_2",
    FilterByTimeStart=60,
    FilterByTimeStop=120
)

# Accept if workspaces differ by one count, but no more
CompareWorkspaces(Workspace1="CG3_4829_1", Workspace2="CG3_4829_2", Tolerance=1)
```
The rest looks :ok_hand: 
@KyleQianliMa  You still need to fix the conda build. Have you already tried?

File: Framework/DataHandling/src/BankPulseTimes.cpp [hunk 1]
```diff
@@ -54,7 +53,13 @@ BankPulseTimes::BankPulseTimes(const std::vector<Mantid::Types::Core::DateAndTim
 BankPulseTimes::BankPulseTimes(::NeXus::File &file, const std::vector<int> &periodNumbers)
     : startTime(DEFAULT_START_TIME), periodNumbers(periodNumbers), have_period_info(true),
       m_sorting_info(PulseSorting::UNKNOWN) {
-  file.openData("event_time_zero");
+
+  // Some old data use "pulse_time" instead of "event_time_zero" as entry
+  try {
+    file.openData("event_time_zero");
+  } catch (std::exception &e) {
```
Comment (peterfpeterson) [inline Framework/DataHandling/src/BankPulseTimes.cpp:60]:
```suggestion
  } catch (const std::exception &) {
```
should fix that pesky windows compiler warning

**Review by jmborr — DISMISSED** [review]

**Comment by KyleQianliMa** [comment]
@jmborr Could you re-run the tests using CG3_4829.nxs.h5 file? It seems like all the check passed somehow. Previously some tests mysteriously failed, so I rebased the repo and pushed again. That caused some problem while I resolve merge conflict. I believe now everything should be ok.

**Review by jmborr — APPROVED** [review]

**Pull request merged**